qt/python/mantidqt/project/recovery/projectrecovery.py

```diff
@@ -0,0 +1,458 @@
+        # respectively. This uses string representation to pass to the C++ algorithm. The outer delimiter is `,`
+        # and the inner delimiter is `+` for the list of lists. e.g. [[a, b],[c, d]] = "a + b , c + d".
+        self.alg_properties_to_ignore = "StartLiveData + PropertyName"
+
+        # The recovery GUI's presenter is set when needed
+        self.recovery_presenter = None
+
+        self.thread_on = False
+
+        # Set to true by workbench on close to kill the thread on completion of project save
+        self.closing_workbench = False
+
+    ######################################################
+    #  Utility
+    ######################################################
+
+    def start_recovery_thread(self):
+        if not self.recovery_enabled:
+            logger.debug("Project Recovery: Recovery thread not started as recovery is disabled")
+            return
+
+        if not self.thread_on:
+            self._timer_thread.start()
+            self.thread_on = True
+
+    def stop_recovery_thread(self):
+        if self._timer_thread is not None:
+            self._timer_thread.cancel()
+            self.thread_on = False
+
+    @staticmethod
+    def _remove_empty_folders_from_dir(directory):
+        folders = glob(os.path.join(directory, "*", ""))
+        for folder in folders:
+            try:
+                os.rmdir(folder)
+            except OSError:
+                # Fail silently as expected for all folders
+                pass
+
+    def _remove_directory_and_directory_trees(self, directory):
+        if directory is None or not os.path.exists(directory):
+            return
+
+        # Only allow deleting in subdirectories of workbench-recovery
+        if self.recovery_directory not in directory:
+            raise RuntimeError("Project Recovery: Only allowed to delete trees in the recovery directory")
+
+        shutil.rmtree(directory)
+
+    @staticmethod
+    def sort_by_last_modified(paths):
+        paths.sort(key=lambda x: os.path.getmtime(x))
+        return paths
+
+    def get_pid_folder_to_be_used_to_load_a_checkpoint_from(self):
+        # Get all pids and order them
+        paths = self.listdir_fullpath(self.recovery_directory_hostname)
+        paths = self.sort_by_last_modified(paths)
+
+        # Get just the pids in a list
+        pids = []
+        for path in paths:
+            pids.append(int(os.path.basename(path)))
+
+        for pid in pids:
+            if not psutil.pid_exists(pid):
+                return os.path.join(self.recovery_directory_hostname, str(pid))
+
+    @staticmethod
+    def listdir_fullpath(directory):
+        try:
+            return [os.path.join(directory, item) for item in os.listdir(directory)]
+        except OSError:
+            return []
+
+    def remove_current_pid_folder(self):
+        self._remove_directory_and_directory_trees(self.recovery_directory_pid)
+
+    ######################################################
```

Comment:
I think it is necessary to have it all in one class, it cuts down on excess referencing in code which accesses this class. It also seems like an unnecessary abstraction layer. So with that in mind yes, it is necessary it helps divide up sections of code meant for specific tasks and is a common practice among programmers.

Reply:
Personally, I would never read that section code, because I navigate via going to definitions and references. However, if other people find it useful then I don't mind it being left in. 

My concern is that it adds another thing to maintain, over time when new functions are added these sections might become misleading. I have seen this many times in our C++ files, which makes files worse than not having sections at all, because reading the comment is not useful.

I still think if you'd like to divide sections of code for specific tasks you should use objects in a OOP language. I think the strongest argument for this, is how much easier it would make the testing of `ProjectRecovery`, as now each component can be tested individually. I can see a drawback of having another reference in calls, but I think that actually improves readability, and certainly making it more testable is worth it.

Reply:
We have discussed this in person and agree that it should stay.

Reply:
There was a slight misunderstanding, I still believe the file should be restructured into smaller classes

Reply:
I don't think it actually makes it more testable. What can be done in separate classes that can't be done in this class? The whole idea of the tests is that they test a unit of the code and each function can be tested as part of this class.

Reply:
Separating them into smaller classes provides a structure and makes it easier to understand and follow. 

Looking at the `projectrecovery` there are 4 _fully_ independent functionalities inside: save, load, checkpoint-repair, and utility.

Moving them into classes greatly reduces the scope of the testing, making it harder to make mistakes, and easier for the next reader to comprehend and follow, as they don't have to keep all of projectrecovery in mind. I also think it would be a lot easier to work on 1 of the 5 100 line files, rather than the current 500-line projectrecovery.

Reply:
I would tend to agree with Dimitar here. Composing classes such as this with other concepts provides clarity in what each level does and allows tests that have less coupling so that fewer tests should break when a given component fails.

As example I would think we are probably missing the concept of a `Checkpoint` class. In my mind a `ProjectRecovery` is a collection of `Checkpoint`s. Removal of them would then be simpler because each `Checkpoint` would know its path and it could have a `.remove` method that does the removal.

Having said that this is probably on the borderline of what I would consider "large".

Reply:
I do not think this class is "Too Large". It serves it's one purpose and it does it in a contained manner yes, it would be possible to refactor all this code to separate classes. I understand that we are diverged and it "shouldn't really be a consideration in this code design" that we keep somewhat true to the legacy project recovery code in MantidPlot. It would only make sense for one person to be responsible for maintaining the MantidPlot/Workbench equivalents of Project Recovery. Unfortunately I won't be here forever and that won't always be me, so we should plan accordingly, making both code bases look similar will make it easier to find your way around, whether you like searching or not. This is a pretty minor reason in the scheme of things and if it provided a genuinely better all round experience I'd put the time into seperating the concepts into classes.

It would however be a bit of a leap into the unknown of how these classes interact and at this stage add bugs that lead to instability.